### PR TITLE
fix(npm): support short sourceUrl format

### PR DESCRIPTION
**Pull request by rarkins**
## Changes

Adds support for npm package source "registry" values being "abc/def", "github:abc/def", "gitlab:abc/def" and "bitbucket:abc/def". Per https://docs.npmjs.com/cli/v8/configuring-npm/package-json#repository

## Context

Closes #17793

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/datasource/npm/get.spec.ts [hunk 1]
```diff
@@ -339,6 +339,46 @@ describe('modules/datasource/npm/get', () => {
...
+    const dep = await getDependency(http, registryUrl, 'vue');
+
+    expect(dep?.sourceUrl).toBe('https://github.com/vuejs/vue');
+    expect(dep?.releases[0].sourceUrl).toBeUndefined();
+    expect(dep?.releases[1].sourceUrl).toBe(
+      'https://github.com/vuejs/vue-next'
+    );
+    expect(dep?.releases[2].sourceUrl).toBe('https://gitlab.com/vuejs/vue');
+    expect(dep?.releases[3].sourceUrl).toBe('https://bitbucket.org/vuejs/vue');
+  });
```
Comment (viceice) [inline lib/modules/datasource/npm/get.spec.ts:380]: why not using a single match object?

File: lib/modules/datasource/npm/get.ts [hunk 2]
```diff
@@ -13,11 +14,31 @@ interface PackageSource {
...
 
+const SHORT_REPO_REGEX = regEx(/^[A-Za-z0-9_.-]+\/[A-Za-z0-9_.-]+$/);
+
 function getPackageSource(repository: any): PackageSource {
   const res: PackageSource = {};
   if (repository) {
     if (is.nonEmptyString(repository)) {
-      res.sourceUrl = repository;
+      // Put github first because then it matches abc/def too
+      if (SHORT_REPO_REGEX.test(repository.replace(/^github:/, ''))) {
```
Comment (viceice) [inline lib/modules/datasource/npm/get.ts:24]: wouldn't it be better to add the optional prefix to the regex and do a single match?

File: lib/modules/datasource/npm/get.ts [hunk 3]
```diff
@@ -13,11 +14,27 @@ interface PackageSource {
   sourceDirectory?: string;
 }
 
+const SHORT_REPO_REGEX = regEx(
+  /^((?<platform>bitbucket|github|gitlab):)?(?<shortRepo>[A-Za-z0-9_.-]+\/[A-Za-z0-9_.-]+)$/
+);
+
+const platformMapping: Record<string, string> = {
+  bitbucket: 'https://bitbucket.org/',
```
[the remark this answers is not in the document]
Comment (rarkins) [reply lib/modules/datasource/npm/get.ts:22]: npm does not document `azure:` as a supported prefix:

image

File: lib/modules/datasource/npm/get.spec.ts [hunk 4]
```diff
@@ -339,6 +339,46 @@ describe('modules/datasource/npm/get', () => {
...
+      .get('/vue')
+      .reply(200, {
+        name: 'vue',
+        repository: {
+          type: 'git',
+          url: 'https://github.com/vuejs/vue',
+        },
+        versions: {
+          '2.0.0': {
+            repository: 'vuejs/vue',
```
[the remark this answers is not in the document]
Comment (rarkins) [reply lib/modules/datasource/npm/get.spec.ts:356]: yes, that would be equivalent

File: lib/modules/datasource/npm/get.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (HonkingGoose) [reply lib/modules/datasource/npm/get.ts:22]: Thanks for the explanation! 😄

File: lib/modules/datasource/npm/get.spec.ts (hunk 1 above)
Comment (viceice) [reply lib/modules/datasource/npm/get.spec.ts:380]: no comment and no code change?

**Review by viceice — COMMENTED** [review]
otherwise LGTM

File: lib/modules/datasource/npm/get.spec.ts (hunk 1 above)
Comment (rarkins) [reply lib/modules/datasource/npm/get.spec.ts:380]: Sorry, misunderstood your question previously. Thought you meant to change the regex to a common one.

Comment (viceice) [reply lib/modules/datasource/npm/get.spec.ts:380]: no problem, i though a single matchObject is easier to read here.

**Review by viceice — APPROVED** [review]

**Pull request merged**